Java: mscorlib types result in duplicates if target assembly is mscorlib.dll

Maybe it is because e4k has to specialize some types in mscorlib.dll, but if I try to bind mscorlib.dll (with a handful of local fixes) it results in the errors below:

		Linking assemblies...
		Error: warning CS1685: The predefined type `System.Object' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.ValueType' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Attribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Int32' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.UInt32' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Int64' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.UInt64' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Single' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Double' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Char' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Int16' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Decimal' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Boolean' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.SByte' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Byte' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.UInt16' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.String' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Enum' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Delegate' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.MulticastDelegate' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Array' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Type' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Collections.IEnumerator' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Collections.IEnumerable' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.IDisposable' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.IntPtr' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.UIntPtr' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.RuntimeFieldHandle' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.RuntimeTypeHandle' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Exception' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.ParamArrayAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.InteropServices.OutAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.ObsoleteAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.InteropServices.DllImportAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.MethodImplAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.InteropServices.MarshalAsAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.InteropServices.InAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.IndexerNameAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Diagnostics.ConditionalAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.CLSCompliantAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Security.Permissions.SecurityAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.RequiredAttributeAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.InteropServices.GuidAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Reflection.AssemblyCultureAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Reflection.AssemblyVersionAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Reflection.AssemblyAlgorithmIdAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Reflection.AssemblyFlagsAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Reflection.AssemblyFileVersionAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Reflection.AssemblyInformationalVersionAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.InteropServices.ComImportAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.InteropServices.CoClassAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.AttributeUsageAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.InteropServices.OptionalAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Security.UnverifiableCodeAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.InteropServices.DefaultCharSetAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.TypeForwardedToAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.FixedBufferAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.CompilerGeneratedAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.InternalsVisibleToAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.RuntimeCompatibilityAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Diagnostics.DebuggerHiddenAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.UnsafeValueTypeAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.InteropServices.UnmanagedFunctionPointerAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Diagnostics.DebuggerBrowsableAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Diagnostics.DebuggerStepThroughAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Diagnostics.DebuggableAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Security.Permissions.HostProtectionAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.ExtensionAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.AsyncStateMachineAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.TupleElementNamesAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.IsReadOnlyAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.IsByRefLikeAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Reflection.DefaultMemberAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.DecimalConstantAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.InteropServices.StructLayoutAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.InteropServices.FieldOffsetAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Reflection.AssemblyProductAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Reflection.AssemblyCompanyAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Reflection.AssemblyCopyrightAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Reflection.AssemblyTrademarkAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.CallerMemberNameAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.CallerLineNumberAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Runtime.CompilerServices.CallerFilePathAttribute' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.TypedReference' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.ArgIterator' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Collections.Generic.IEnumerable' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Collections.Generic.IList' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Collections.Generic.IReadOnlyList' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Collections.Generic.ICollection' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Collections.Generic.IReadOnlyCollection' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Nullable' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Threading.Tasks.Task' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.Threading.Tasks.Task' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.IFormattable' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.FormattableString' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.ValueTuple' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.ValueTuple' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.ValueTuple' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.ValueTuple' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.ValueTuple' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.ValueTuple' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.ValueTuple' is defined multiple times. Using definition from `mscorlib.dll'
		Error: warning CS1685: The predefined type `System.ValueTuple' is defined multiple times. Using definition from `mscorlib.dll'
		Error: /tmp/x94te73t.0.cs(21,24) : error CS0433: The imported type `System.Int32' is defined multiple times
		Error: /tmp/x94te73t.0.cs(21,43) : error CS0433: The imported type `System.IntPtr' is defined multiple times
		Error: /tmp/x94te73t.0.cs(21,53) : error CS0433: The imported type `System.String' is defined multiple times

	Unhandled Exception:
	System.Exception: Resource.designer.dll compilation failed! See e4k/android/Resource.designer.cs for details.
	  at Embeddinator.Driver.CreateAar () [0x00587] in /sources/Embeddinator-4000/binder/Compilation.cs:494 
	  at Embeddinator.Driver.CompileCode () [0x00167] in /sources/Embeddinator-4000/binder/Compilation.cs:68 
	  at Embeddinator.Driver.Run () [0x000d4] in /sources/Embeddinator-4000/binder/Driver.cs:253 
	  at Embeddinator.CLI.Main (System.String[] args) [0x000a6] in /sources/Embeddinator-4000/binder/CLI.cs:220 
	[ERROR] FATAL UNHANDLED EXCEPTION: System.Exception: Resource.designer.dll compilation failed! See e4k/android/Resource.designer.cs for details.
	  at Embeddinator.Driver.CreateAar () [0x00587] in /sources/Embeddinator-4000/binder/Compilation.cs:494 
	  at Embeddinator.Driver.CompileCode () [0x00167] in /sources/Embeddinator-4000/binder/Compilation.cs:68 
	  at Embeddinator.Driver.Run () [0x000d4] in /sources/Embeddinator-4000/binder/Driver.cs:253 
	  at Embeddinator.CLI.Main (System.String[] args) [0x000a6] in /sources/Embeddinator-4000/binder/CLI.cs:220 

On the other hand, I could successfully bind System.dll (to the point that "System.aar" is generated), so this issue is specific to mscorlib.dll as the target.

--- a/binder/Utils/ResourceDesignerGenerator.cs
+++ b/binder/Utils/ResourceDesignerGenerator.cs
@@ -22,6 +22,7 @@ namespace Embeddinator
         readonly CodeCompileUnit unit = new CodeCompileUnit();
         readonly CSharpCodeProvider csc = new CSharpCodeProvider();
         readonly CodeGeneratorOptions options = new CodeGeneratorOptions();
+        readonly List<IKVM.Reflection.Assembly> usedAssemblies = new List<IKVM.Reflection.Assembly>();
 
         public IList<IKVM.Reflection.Assembly> Assemblies { get; set; }
 
@@ -77,7 +78,6 @@ namespace Embeddinator
                 new CodeAttributeArgument("IsApplication", new CodeSnippetExpression("true"))));
 
             var ns = new CodeNamespace("__embeddinator__");
-            ns.Imports.Add(new CodeNamespaceImport("System"));
             ns.Imports.Add(new CodeNamespaceImport("Android.Runtime"));
             unit.Namespaces.Add(ns);
 
@@ -91,18 +91,21 @@ namespace Embeddinator
                 new CodeAttributeArgument(new CodeSnippetExpression("\"Xamarin.Android.Build.Tasks\"")),
                 new CodeAttributeArgument(new CodeSnippetExpression("\"1.0.0.0\""))));
 
+            var intPtr = new CodeTypeReference(typeof(IntPtr));
+
             var readFieldInt = new CodeMemberMethod
             {
                 Name = "ReadFieldInt",
                 Attributes = MemberAttributes.Private | MemberAttributes.Static,
-                ReturnType = new CodeTypeReference("Int32"),
+                ReturnType = new CodeTypeReference(typeof(int)),
             };
-            readFieldInt.Parameters.Add(new CodeParameterDeclarationExpression("IntPtr", "R"));
-            readFieldInt.Parameters.Add(new CodeParameterDeclarationExpression("String", "fieldName"));
+            readFieldInt.Parameters.Add(new CodeParameterDeclarationExpression(intPtr, "R"));
+            readFieldInt.Parameters.Add(new CodeParameterDeclarationExpression(typeof(string), "fieldName"));
             resource.Members.Add(readFieldInt);
 
+            readFieldInt.Statements.Add(new CodeVariableDeclarationStatement(intPtr, "fieldId"));
             readFieldInt.Statements.Add(new CodeAssignStatement(
-                new CodeSnippetExpression("IntPtr fieldId"),
+                new CodeVariableReferenceExpression("fieldId"),
                 new CodeSnippetExpression("JNIEnv.GetStaticFieldID(R, fieldName, \"I\")")));
             readFieldInt.Statements.Add(new CodeMethodReturnStatement(new CodeSnippetExpression("JNIEnv.GetStaticIntField(R, fieldId)")));
 
@@ -110,19 +113,21 @@ namespace Embeddinator
             {
                 Name = "ReadFieldArray",
                 Attributes = MemberAttributes.Private | MemberAttributes.Static,
-                ReturnType = new CodeTypeReference("Int32[]"),
+                ReturnType = new CodeTypeReference(typeof(int[])),
             };
-            readFieldArray.Parameters.Add(new CodeParameterDeclarationExpression("IntPtr", "R"));
-            readFieldArray.Parameters.Add(new CodeParameterDeclarationExpression("String", "fieldName"));
+            readFieldArray.Parameters.Add(new CodeParameterDeclarationExpression(intPtr, "R"));
+            readFieldArray.Parameters.Add(new CodeParameterDeclarationExpression(typeof(string), "fieldName"));
             resource.Members.Add(readFieldArray);
 
+            readFieldArray.Statements.Add(new CodeVariableDeclarationStatement(intPtr, "fieldId"));
+            readFieldArray.Statements.Add(new CodeVariableDeclarationStatement(intPtr, "value"));
             readFieldArray.Statements.Add(new CodeAssignStatement(
-                new CodeSnippetExpression("IntPtr fieldId"),
+                new CodeVariableReferenceExpression("fieldId"),
                 new CodeSnippetExpression("JNIEnv.GetStaticFieldID(R, fieldName, \"[I\")")));
             readFieldArray.Statements.Add(new CodeAssignStatement(
-                new CodeSnippetExpression("IntPtr value"),
+                new CodeVariableReferenceExpression("value"),
                 new CodeSnippetExpression("JNIEnv.GetStaticObjectField(R, fieldId)")));
-            readFieldArray.Statements.Add(new CodeMethodReturnStatement(new CodeSnippetExpression("JNIEnv.GetArray<Int32>(value)")));
+            readFieldArray.Statements.Add(new CodeMethodReturnStatement(new CodeSnippetExpression("JNIEnv.GetArray<int>(value)")));
 
             var updateIdValues = new CodeMemberMethod
             {
@@ -131,7 +136,7 @@ namespace Embeddinator
             };
             resource.Members.Add(updateIdValues);
 
-            updateIdValues.Statements.Add(new CodeVariableDeclarationStatement("IntPtr", "R"));
+            updateIdValues.Statements.Add(new CodeVariableDeclarationStatement(intPtr, "R"));
 
             foreach (var assembly in Assemblies)
             {
@@ -208,6 +213,8 @@ namespace Embeddinator
                                         throw new Exception($"Type {field.FieldType.FullName} from member {nested.FullName}.{field.Name} not supported for Resource fields!");
                                     }
                                     updateIdValues.Statements.Add(new CodeAssignStatement(left, right));
+                                    if (!usedAssemblies.Contains(assembly))
+                                        usedAssemblies.Add(assembly);
                                 }
                             }
                         }
@@ -229,7 +236,7 @@ namespace Embeddinator
             parameters.ReferencedAssemblies.Add(XamarinAndroid.FindAssembly("System.Runtime.dll"));
             parameters.ReferencedAssemblies.Add(XamarinAndroid.FindAssembly("Java.Interop.dll"));
             parameters.ReferencedAssemblies.Add(XamarinAndroid.FindAssembly("Mono.Android.dll"));
-            foreach (var assembly in Assemblies)
+            foreach (var assembly in usedAssemblies)
             {
                 parameters.ReferencedAssemblies.Add(assembly.Location);
             }

--- a/tests/MonoEmbeddinator4000.Tests/Approvals/ResourceDesignerTest.Anim.approved.txt
+++ b/tests/MonoEmbeddinator4000.Tests/Approvals/ResourceDesignerTest.Anim.approved.txt
@@ -11,26 +11,28 @@
 [assembly: Android.Runtime.ResourceDesignerAttribute("__embeddinator__.Resource", IsApplication=true)]
 
 namespace @__embeddinator__ {
-    using System;
     using Android.Runtime;
     
     
     [System.CodeDom.Compiler.GeneratedCodeAttribute("Xamarin.Android.Build.Tasks", "1.0.0.0")]
     public class Resource {
         
-        private static Int32 ReadFieldInt(IntPtr R, String fieldName) {
-            IntPtr fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "I");
+        private static int ReadFieldInt(System.IntPtr R, string fieldName) {
+            System.IntPtr fieldId;
+            fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "I");
             return JNIEnv.GetStaticIntField(R, fieldId);
         }
         
-        private static Int32[] ReadFieldArray(IntPtr R, String fieldName) {
-            IntPtr fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "[I");
-            IntPtr value = JNIEnv.GetStaticObjectField(R, fieldId);
-            return JNIEnv.GetArray<Int32>(value);
+        private static int[] ReadFieldArray(System.IntPtr R, string fieldName) {
+            System.IntPtr fieldId;
+            System.IntPtr value;
+            fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "[I");
+            value = JNIEnv.GetStaticObjectField(R, fieldId);
+            return JNIEnv.GetArray<int>(value);
         }
         
         public static void UpdateIdValues() {
-            IntPtr R;
+            System.IntPtr R;
             R = JNIEnv.FindClass("com.mono.embeddinator.R$anim");
             managedandroid.Resource.Animation.THIS_IS_CAPS = ReadFieldInt(R, "THIS_IS_CAPS");
             managedandroid.Resource.Animation.applicationName = ReadFieldInt(R, "applicationName");

--- a/tests/MonoEmbeddinator4000.Tests/Approvals/ResourceDesignerTest.Full.approved.txt
+++ b/tests/MonoEmbeddinator4000.Tests/Approvals/ResourceDesignerTest.Full.approved.txt
@@ -11,26 +11,28 @@
 [assembly: Android.Runtime.ResourceDesignerAttribute("__embeddinator__.Resource", IsApplication=true)]
 
 namespace @__embeddinator__ {
-    using System;
     using Android.Runtime;
     
     
     [System.CodeDom.Compiler.GeneratedCodeAttribute("Xamarin.Android.Build.Tasks", "1.0.0.0")]
     public class Resource {
         
-        private static Int32 ReadFieldInt(IntPtr R, String fieldName) {
-            IntPtr fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "I");
+        private static int ReadFieldInt(System.IntPtr R, string fieldName) {
+            System.IntPtr fieldId;
+            fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "I");
             return JNIEnv.GetStaticIntField(R, fieldId);
         }
         
-        private static Int32[] ReadFieldArray(IntPtr R, String fieldName) {
-            IntPtr fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "[I");
-            IntPtr value = JNIEnv.GetStaticObjectField(R, fieldId);
-            return JNIEnv.GetArray<Int32>(value);
+        private static int[] ReadFieldArray(System.IntPtr R, string fieldName) {
+            System.IntPtr fieldId;
+            System.IntPtr value;
+            fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "[I");
+            value = JNIEnv.GetStaticObjectField(R, fieldId);
+            return JNIEnv.GetArray<int>(value);
         }
         
         public static void UpdateIdValues() {
-            IntPtr R;
+            System.IntPtr R;
             R = JNIEnv.FindClass("com.mono.embeddinator.R$attr");
             managedandroid.Resource.Attribute.hello = ReadFieldInt(R, "hello");
             R = JNIEnv.FindClass("com.mono.embeddinator.R$id");

--- a/tests/MonoEmbeddinator4000.Tests/Approvals/ResourceDesignerTest.FullWithResourceFile.approved.txt
+++ b/tests/MonoEmbeddinator4000.Tests/Approvals/ResourceDesignerTest.FullWithResourceFile.approved.txt
@@ -11,26 +11,28 @@
 [assembly: Android.Runtime.ResourceDesignerAttribute("__embeddinator__.Resource", IsApplication=true)]
 
 namespace @__embeddinator__ {
-    using System;
     using Android.Runtime;
     
     
     [System.CodeDom.Compiler.GeneratedCodeAttribute("Xamarin.Android.Build.Tasks", "1.0.0.0")]
     public class Resource {
         
-        private static Int32 ReadFieldInt(IntPtr R, String fieldName) {
-            IntPtr fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "I");
+        private static int ReadFieldInt(System.IntPtr R, string fieldName) {
+            System.IntPtr fieldId;
+            fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "I");
             return JNIEnv.GetStaticIntField(R, fieldId);
         }
         
-        private static Int32[] ReadFieldArray(IntPtr R, String fieldName) {
-            IntPtr fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "[I");
-            IntPtr value = JNIEnv.GetStaticObjectField(R, fieldId);
-            return JNIEnv.GetArray<Int32>(value);
+        private static int[] ReadFieldArray(System.IntPtr R, string fieldName) {
+            System.IntPtr fieldId;
+            System.IntPtr value;
+            fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "[I");
+            value = JNIEnv.GetStaticObjectField(R, fieldId);
+            return JNIEnv.GetArray<int>(value);
         }
         
         public static void UpdateIdValues() {
-            IntPtr R;
+            System.IntPtr R;
             R = JNIEnv.FindClass("com.mono.embeddinator.R$attr");
             managedandroid.Resource.Attribute.hello = ReadFieldInt(R, "hello");
             R = JNIEnv.FindClass("com.mono.embeddinator.R$id");

--- a/tests/MonoEmbeddinator4000.Tests/Approvals/ResourceDesignerTest.String.approved.txt
+++ b/tests/MonoEmbeddinator4000.Tests/Approvals/ResourceDesignerTest.String.approved.txt
@@ -11,26 +11,28 @@
 [assembly: Android.Runtime.ResourceDesignerAttribute("__embeddinator__.Resource", IsApplication=true)]
 
 namespace @__embeddinator__ {
-    using System;
     using Android.Runtime;
     
     
     [System.CodeDom.Compiler.GeneratedCodeAttribute("Xamarin.Android.Build.Tasks", "1.0.0.0")]
     public class Resource {
         
-        private static Int32 ReadFieldInt(IntPtr R, String fieldName) {
-            IntPtr fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "I");
+        private static int ReadFieldInt(System.IntPtr R, string fieldName) {
+            System.IntPtr fieldId;
+            fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "I");
             return JNIEnv.GetStaticIntField(R, fieldId);
         }
         
-        private static Int32[] ReadFieldArray(IntPtr R, String fieldName) {
-            IntPtr fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "[I");
-            IntPtr value = JNIEnv.GetStaticObjectField(R, fieldId);
-            return JNIEnv.GetArray<Int32>(value);
+        private static int[] ReadFieldArray(System.IntPtr R, string fieldName) {
+            System.IntPtr fieldId;
+            System.IntPtr value;
+            fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "[I");
+            value = JNIEnv.GetStaticObjectField(R, fieldId);
+            return JNIEnv.GetArray<int>(value);
         }
         
         public static void UpdateIdValues() {
-            IntPtr R;
+            System.IntPtr R;
             R = JNIEnv.FindClass("com.mono.embeddinator.R$string");
             managedandroid.Resource.String.THIS_IS_CAPS = ReadFieldInt(R, "THIS_IS_CAPS");
             managedandroid.Resource.String.applicationName = ReadFieldInt(R, "applicationName");

--- a/tests/MonoEmbeddinator4000.Tests/DriverTest.cs
+++ b/tests/MonoEmbeddinator4000.Tests/DriverTest.cs
@@ -209,6 +209,16 @@ namespace Embeddinator.Tests
             RunDriver("Interfaces");
         }
 
+        [Test, Category ("Slow")]
+        public void DuplicateMscorlibTypes()
+        {
+            options.Compilation.Platform = TargetPlatform.Android;
+            options.GeneratorKind = GeneratorKind.C;
+            RunDriver("mscorlib");
+            options.GeneratorKind = GeneratorKind.Java;
+            RunDriver("mscorlib");
+        }
+
         /// <summary>
         /// Validates we get native libraries from the assembly
         /// </summary>

--- a/tests/MonoEmbeddinator4000.Tests/MonoEmbeddinator4000.Tests.csproj
+++ b/tests/MonoEmbeddinator4000.Tests/MonoEmbeddinator4000.Tests.csproj
@@ -100,6 +100,7 @@
     <EmbeddedResource Include="Samples\Enums.cs" />
     <EmbeddedResource Include="Samples\Interfaces.cs" />
     <EmbeddedResource Include="Samples\Resource.Anim.cs" />
+    <EmbeddedResource Include="Samples\mscorlib.cs" />
   </ItemGroup>
   <Import Project="$(MSBuildBinPath)\Microsoft.CSharp.targets" />
 </Project>

--- a/tests/MonoEmbeddinator4000.Tests/Samples/mscorlib.cs
+++ b/tests/MonoEmbeddinator4000.Tests/Samples/mscorlib.cs
@@ -1,0 +1,8 @@
+﻿namespace System
+{
+    public class String { }
+
+    public class Int32 { }
+
+    public class IntPtr { }
+}

